cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,277 @@
+  b. The IPCache determines the IP's full set of labels, allocates an identity, and informs the SelectorCache of this new IP
+2. The selector is updated:
+  a. The NameManager calls `sc.UpdateFQDNSelector()`, updating the set of selected IPs for this given FQDNSelector
+  b. The SelectorCache determines the complete set of identities selected
+  c. Any differences are pushed to network policies
+3. Endpoints are updated
+  Any pending incremental updates are batched up and applied to the endpoints' BPF PolicyMap. If needed, Envoy is also updated
+ 
+Only at this point can the DNS packet be returned to the waiting endpoint.
+
+## Performance bottlenecks
+
+### Many identities
+
+Some DNS names (e.g S3) will return many different IP addresses. S3 even has a TTL of 5 seconds, so the set of IPs will quickly be very broad. This causes a large number of identities to be allocated -- one per IP. These identities then need to plumbed through to all endpoints.
+
+### Label sizes
+
+CIDR labels, especially IPv6 labels, are large. For example, the full set of labels for the CIDR `2001:db8:cafe::cab:4:b0b:0/112` is 2KiB. Every extra-cluster identity always has the full set of CIDR identities calculated for it.
+
+----
+
+## Proposal
+
+This CFP proposes 3 key changes. Together, they preserve correctness while radically restructuring how prefixes are selected.
+
+1: Labels on prefixes propagate downwards. That is to say, if prefix `192.0.0.0/22` has label `foo:bar`, then prefix `192.0.2.0/24` also has label `foo:bar`, unless overridden. The IPCache aggregates labels accordingly.
+2: Only selected CIDR labels are inserted in to the ipcache. If a selector references CIDR `192.0.2.0/24`, only the prefix -> label mapping `192.0.2.0/24 -> cidr:192.0.2.0/24` is added to the IPCache.
+3: When the NameManager learns about a new relevant `name -> IP` pair, it upserts the label `IP -> (dns:selector)` It **does not** create any CIDR labels.
+
+
+### 1: CIDR Label down-propagation
+
+Currently, prefixes in the ipcache are independent. Instead, labels should flow downwards from prefixes. For example, consider an IPCache with two prefixes: 
+
+```
+192.0.2.0/24 -> k8s:foo:bar, cidr:192.0.2.0/24
+192.0.2.2/32 -> cidr:192.0.2.2/32, reserved:kube-apiserver
+```
+
+Then, the complete set of labels for `192.0.2.2/32` would be `k8s:foo:bar, cidr:192.0.2.0/24, cidr:192.0.2.2/32, reserved:kube-apiserver`.
+
+This is a cornerstone for implementing another feature, non-k8s label sources. It is also required for correctness within the context of this proposal, especially when ToFQDN and CIDR selectors overlap.
+
+#### Lifecycle
+
+With label down-propagation, it is important that updates are propagated accordingly. When a prefix P is updated or deleted, then updates must be also triggered for all prefixes contained in P.
+
+This update may be calculated if we use a bitwise LPM trie. Alternatively, scanning the entire ipcache to determine affected prefixes may be more efficient.
```

Comment:
Why would scanning the entire ipcache be more efficient than doing a descendents iterator?

Reply:
I prototyped this in a different context, and discovered that trie iteration causes heap allocation, whereas a full map scan did not. In my benchmark, even though I reduced regular expression evaluations by a factor of 10, it was a 20% slow-down once GC time was added in.

Since scanning a small-ish map and checking for prefix coverage is alloc-less and cheap, I would expect it to outperform more complex data structures until map sizes grow very large.

In any case, this will be very easy to test.

Reply:
update: this has currently merged with simple scanning. If we need to move this to prefix-aware selection, we can do so as needed.